README.md

[ ![PomodoroTasks2 in copr](https://copr.fedorainfracloud.org/coprs/liloman/githubs/package/pomodoroTasks2/status_image/last_build.png "PomodoroTasks2 in copr")](https://copr.fedorainfracloud.org/coprs/liloman/githubs/package/pomodoroTasks2)

Don't make any excuse anymore to not use the [Pomodoro Technique wikipedia](https://en.wikipedia.org/wiki/Pomodoro_Technique) or [The Pomodoro Technique ](extras/technique.pdf)!

Pomodoro technique allows you to concentrate on the current task and take short breaks meanwhile works.
If you get that and join it with a task manager alike taskwarrior (or any other) you can have a complete workflow, accounting the time spend on any task meanwhile you take the proper rests for your brain, body, life and eyes. :)

Table of Contents
=================

   * [Table of Contents](#table-of-contents)
   * [INSTALL](#install)
      * [Packages](#packages)
      * [Manual](#manual)
   * [Why do I need timewarrior?](#why-do-i-need-timewarrior)
   * [Reminders](#reminders)
   * [Screenshots](#screenshots)
   * [Spec](#spec)
   * [TODO](#todo)
   * [FIXED](#fixed)

INSTALL
=================

Packages
--------------

1. Fedora 24/25 x86:

```bash
dnf copr enable liloman/githubs
dnf install pomodoroTasks2
```

This will install all the timewarrior stuff and set the enviroment properly.

Manual
--------------

1. Taskwarrior dependencies (python based)

```bash
pip install tasklib --user
sudo dnf/apt-get/whatever install taskwarrior/task/whatever
task <<< yes
```

2. Timewarrior

```bash
sudo dnf/apt-get/whatever install build-essential cmake 
git clone --recursive https://git.tasktools.org/scm/tm/timew.git timew.git
cd timew.git
git checkout master 
cmake -DCMAKE_BUILD_TYPE=release .
make
sudo make install
timew <<< yes
```

3. PomodoroTasks2

```bash
git clone https://github.com/liloman/pomodoroTasks2
cd pomodoroTasks2/
./pomodoro-daemon.py
```

You can customize the working time and the break times (short and long), just exporting a few ENV variables in your ~/.bashrc.

```bash
#default pomodoro session (minutes)
export POMODORO_TIMEOUT=25
#default pomodoro short break (minutes)
export POMODORO_STIMEOUT=5
#default pomodoro long break (minutes)
export POMODORO_LTIMEOUT=15
```


So just launch the pomodoro-daemon.py and you are ready to go, feel free to add it in ~/.local/bin,autostart,systemd,... :)


Why do I need timewarrior?
=================

Because you the objective is track all your workflow and nothing better for that purpose than the newcomer and taskwarrior brother timewarrior. :)

If you wish to track every task of taskwarrior in timewarrior you need to:

1. Execute the extras/prepare_hooks.sh script:

 So it will be: 

 ```bash
 git clone https://github.com/liloman/pomodoroTasks2
 cd pomodoroTasks2/
 ./extras/prepare_hooks.sh install .
 ```

 And for now on: 
 a. Each time you start/stop a task it will be tracked with timewarrior unless it contains the +notimew tag. ;) 

 b. Timewarrior will stop tracking on logout/shutdown

 c. Everytime that there's a pomodoro timeout the app automatically track  it

 d. You have a new timewarrior report for your work (everything within a +nowork tag) ;)


2. You can create these two aliases for cozyness in your ~/.bashrc I can guarantee that you will use them ;)
 
 
 ```bash
 alias twt='timew work today'
 alias tww='timew work :week'
 ```

3. If you want to track also the time the PC is off (I don't recommend it if you have several PCs and sync your tasks among them) you can execute after every log [this script](https://github.com/liloman/warriors/blob/master/last-boots.sh)

 So it will be: 
 
 ```bash
 cd your-autostart-dir/xinit/systemd/whatever/...
 wget https://raw.githubusercontent.com/liloman/warriors/master/last-boots.sh
 chmod u+x last-boots.sh
 
 ```
 
4. Start enjoying timewarrior!. :) 


 How much have I been working today? 
 
 ```bash
 prompt>twt
 Total by Tag from 2017-01-15 23:00:00 - 2017-01-16 09:59:28 (sorted by time)
 
 Tag                                                Total
  ----------------------------------       ---------------
 
 More stuff / pro:stuff                   0 days, 2:14:00
 Doing demo / pro:dev.demo                0 days, 3:04:49
                                         ----------------                   
 Total                                    0 days, 5:18:49
 
 ```

 Umm, 5:18 hours. Niiiiiiiiiiice. 
 
 So no remorse of conscience for today. ;)
 
 Of course you could know how much have you worked this week for a project for example:
 
 ```bash
  prompt>tww pro:awesome
 Total by Tag from 2017-01-08 23:00:00 - 2017-01-15 23:00:00 (sorted by time)
 
 Tag                                                Total
 ----------------------------------       ---------------
 
 Fixing that bugs 1/ pro:awesome          0 days, 0:14:00
 Fixing that bugs 2 / pro:awesome         1 days, 2:14:00
 Writing docs / +docs / pro:awesome       3 days, 3:04:49
                                         ----------------                   
 Total                                    4 days, 5:32:49
 
 ```

Reminders 
=================

It's another functionality included into pomodoroTasks2.


My use case is for that kind of tasks that you need to be reminded about and most time even on advance.  So I use it to remember birthdays, due times or important tasks for this week in conclusión as a tasks calendar. :)

It's based on taskwarrior recurrence tasks and basically you just need to create a task with the +reminder tag and it will be displayed on timeout as a remind for your rewarded spare time. ;)

I include the script I use myself to create them or you can use whatever you want just include a +reminder tag on any task and it'll be show up on timeout if it's on pending status.

My script can create 3 types of reminders:

1. Recurrent (ex: birthdate): You create a birthdate and set to be shown with x days on advance until its due date

2. NonRecurrent(from date): Create a task for the day/month/year and set it to be show x days on advance until its due date

3. Inmediate(from today!): Create a task to be shown from now on or after x days until it's done


If you installed the package it will add a new command to /usr/bin/ otherwise you have to copy it like:

 ```bash
git clone https://github.com/liloman/pomodoroTasks2
cd pomodoroTasks2/extras
cp add-reminder.sh wherever-you-want-it
 ```

Workflow to create the example:

 ```bash
$add-reminder.sh
Select the type of reminder
1) Recurrent (ex: birthdate)  3) Inmediate(from today!)
2) NonRecurrent(from date)
#? 3
Description:Example reminder for screenshot!
task add 'Example reminder for screenshot!' pro:tasks due:someday wait:today until:due+1d recur: +reminder  rc.dateformat=D/M/Y
Are you sure?
1) Yes
2) No
#? 1
Created task 35.
 ```

And here it's the result:

![Showing reminder](images/screenshots/reminder.png "Showing reminder")

When you mark the task as done or check the checkbox it will be disappear from the timeout splash screen.


Screenshots
=================

Relax time:

![25 minutes passed](images/screenshots/timer1.png "25 minutes passed")

Back to work:

![Back to work?](images/screenshots/timer2.png "Back to work?")

Menu:

![Started with tooltip](images/screenshots/started.png "Started with tooltip")

![Paused with menu](images/screenshots/paused.png "Paused with menu")

All trayicons:

![All icons](images/screenshots/all-icons.png "All icons")

Change current task:

![Change Task](images/screenshots/changeTask.png "Change task")

![Change Task 2](images/screenshots/changeTask2.png "Change task 2")

Add new task:

![Add new Task](images/screenshots/addTask.png "Add new Task")



Spec
=================

Minimalistic implementation with FSM (Finite State Machine) and some dbus niceness. :)


TODO
=================

- [x] dbus 
- [x] Timewarrior integration to track the complete lifespan of your computer. ;)
- [x] Dont wait a minute to refresh the trayicon when used the cli
- [x] Script to install/uinstall hooks
- [x] Explain/include reminders
- [ ] Add packages for Ubuntu/Debian/Arch/...
- [ ] Unit testing \(fix travis install ... \)

FIXED
=================

1. ~~Linux mint issues~~
2. ~~Eliminated cycle when closed app from the sysicon~~


CHANGELOG
0.X
*Arch packages
*Improved UX
*Moved to classes 
*Moved to python 3
-Use setup.py

0.4
*Fix tests
*Autostart user unit on login

0.3
*Full dnf/rpm install with hooks

0.2
*Client with autocompletion


0.1 
*Integration with taskwarrior
*Integration with timewarrior tracking every second (see https://github.com/liloman/warriors)
*moved to python
*dbus
*testing


pomodorotasks/daemon.py
#!/usr/bin/env python
# Pomodoro daemon with FSM

from __future__ import print_function
from __future__ import division
from builtins import str
from past.utils import old_div
import os
import threading
from tasklib import TaskWarrior, Task
# import datetime 
from datetime import datetime

#For GUI 
import subprocess
from gi.repository import GObject
import sys

# For dbus
import dbus.service
import dbus.exceptions
import dbus.mainloop.glib

dirname, filename = os.path.split(os.path.realpath(__file__))
#change to the project real dir
os.chdir(dirname)

class Pomodoro(dbus.service.Object):

    #Finite State Machine logic (FSM)
    # [from-event]=to
    events = {
            'started-stop': 'do_stop',
            'started-start': 'do_warning' ,
            'started-pause': 'do_pause' ,
            'started-stop': 'do_stop',
            'started-reset': 'do_reset',
            'started-status': 'do_status',
            'paused-start': 'do_resume',
            'paused-stop': 'do_stop',
            'paused-reset': 'do_reset',
            'paused-pause': 'do_resume',
            'paused-status': 'do_status',
            'stopped-start': 'do_resume',
            'stopped-pause': 'do_resume',
            'stopped-stop': 'do_warning',
            'stopped-reset': 'do_reset',
            'stopped-status': 'do_status',
            'started-take_break': 'do_timeout',
            'started-dry_start': 'do_warning' ,
            'started-dry_stop': 'do_dry_stop',
            'stopped-dry_start': 'do_dry_start',
            'paused-dry_start': 'do_dry_start',
            'stopped-dry_stop': 'do_warning',
            }
    bus_name = "org.liloman.pomodoro"
    state = ""
    #timeout pomodoro (minutes)
    timer_pomodoro = int(os.getenv('POMODORO_TIMEOUT',25))
    #pomodoro short break time (minutes)
    stimeout = str(os.getenv('POMODORO_STIMEOUT',5))
    #pomodoro long break time (minutes)
    ltimeout = str(os.getenv('POMODORO_LTIMEOUT',15))
    #Number of breaks to take a long break (LONG_TIME_BREAK)
    maxbreaks = 4
    #timeout wait for events (seconds)
    timeout = 60
    #Counter for breaks
    breaks = 1
    #Continuos mode
    continuous = False
    #Total time elapsed
    time_elapsed = 0
    last_task_id = 0
    tw = None 
    timer = None 
    # default systray icon
    systrayIcon = "images/iconStarted-0.png"

    def __init__(self, rc = '~/.task'):
        self.tw = TaskWarrior(data_location=rc, create=True)
        try:
            name = dbus.service.BusName(self.bus_name, bus=dbus.SessionBus(),do_not_queue=True, replace_existing=False, allow_replacement=False )
            dbus.service.Object.__init__(self, name, '/daemon')
        except dbus.exceptions.NameExistsException:
            print("Daemon is already running.")
            sys.exit(0)

    #get active task
    def get_active_task(self):
        for task in self.tw.tasks.pending():
            if task.active:
                #get all fields in task
                task.refresh()
                return task
        return {}

    def update_systray(self):
        try:
            systray_bus = dbus.SessionBus()
            systray_bus = systray_bus.get_object('org.liloman.pomodoro.systray', "/systray")
            interface_systray = dbus.Interface(systray_bus, "org.liloman.pomodoro.systrayInterface")
            status = self.do_status()
            interface_systray.set_state(dbus.Dictionary({'icon': self.systrayIcon, 'state': self.state, 'tooltip': status }, signature='ss'))
        except:
            #print "not systray found"
            return

    def do_pause(self):
        active = self.get_active_task ()
        #if the user mark current as done and dont select a new task to work in
        if active:
            self.last_task_id = active['uuid']
            active.stop ()
        self.state = "paused"
        return "ok"

    #Resume the last task for previous pause/stop
    def do_resume(self):
        #If resumed from pause/stop without changing the current task from GUI
        if  self.last_task_id != 0:
            try:
                new=self.tw.tasks.pending().get(uuid=self.last_task_id)
                new.start()
            except: #try to resume a now done/deleted/... task
                self.last_task_id = 0
        else: #let just use the timer without any task working in
            self.state = "started"
            return "no previous task"

        #only update time_elapsed when paused
        if  self.state == "stopped":  
            self.time_elapsed = 0 
        self.state = "started"

        return "ok"

    def do_stop(self):
        active = self.get_active_task ()
        if active:
            self.last_task_id = active['uuid']
            active.stop()
        self.state = "stopped"
        self.time_elapsed = 0
        return "ok"

    def do_reset(self):
        self.do_stop()
        self.breaks = 1
        if self.last_task_id != 0:
            self.do_start(dbus.Dictionary({'uuid': self.last_task_id, 'resume': 'No'}))
        else: # if the user wants to reset the timer with no task working in
            self.state = "started"
            return "no previous task"

        return "ok"

    def do_warning(self):
        return "Already "+self.state

    def do_status(self):
        remaining = "%02d:%02d" % divmod(self.timer_pomodoro*60 - self.time_elapsed,60)
        active = self.get_active_task ()
        msg = ""
        project = "None"
        desc = "None"
        if active:
            self.state = "started"
            if active['project']:
                #must work in utf8
                if u''.join(active['project']).encode('utf-8').strip()  != "":
                    project=u''.join(active['project']).encode('utf-8').strip() 
            if active['description']:
                if u''.join(active['description']).encode('utf-8').strip()  != "":
                    desc=u''.join(active['description']).encode('utf-8').strip() 
            msg="\nBreak num:"+str(self.breaks)+" Continuous mode:"+str(self.continuous)+"\nProject:"+project+"\n"+desc
        elif self.last_task_id != 0:
            last=self.tw.tasks.get(uuid=self.last_task_id)
            last.refresh()
            if last['project']:
                #must work in utf8
                if u''.join(last['project']).encode('utf-8').strip()  != "":
                    project=u''.join(last['project']).encode('utf-8').strip() 
            if last['description']:
                if u''.join(last['description']).encode('utf-8').strip()  != "":
                    desc=u''.join(last['description']).encode('utf-8').strip() 
            msg="\nBreak num:"+str(self.breaks)+" Continuous mode:"+str(self.continuous)+"\nLast Project:"+project+"\n"+desc

        rest = old_div(self.timer_pomodoro, 8)
        if rest > 0:
            actual=old_div(self.time_elapsed, (rest*60))
        else:
            actual=0
        i_started = "images/iconStarted-{0}.png".format(actual)
        i_no_task = "images/iconStarted-no-task.png"
        i_paused  = "images/iconPaused.png"
        i_stopped = "images/iconStopped.png"
        if self.state == "started" and not active:
            self.systrayIcon=i_no_task 
        elif self.state == "started":
            self.systrayIcon=i_started 
        elif self.state == "paused":
            self.systrayIcon=i_paused
        elif self.state == "stopped":
            self.systrayIcon=i_stopped 

        return self.state+" "+remaining+" left"+msg

    def do_timeout(self):
        if self.continuous:
           self.time_elapsed = 0
           if self.breaks == self.maxbreaks:
               self.breaks = 0
           self.breaks+=1
           return "call do_timeout.py"
        self.do_stop()
        if self.breaks == self.maxbreaks:
            cmd = ['./do_timeout.py', self.ltimeout, str(self.breaks) ]
            # reset number of breaks
            self.breaks = 0
        else:
            cmd = ['./do_timeout.py', self.stimeout, str(self.breaks) ]
        subprocess.Popen(cmd)
        self.breaks+=1
        return "call do_timeout.py"

    def increment(self):
        self.timer = threading.Timer(5.0, self.increment)
        self.timer.start()
        self.update_systray()
        if self.state == "started":
            self.time_elapsed+=5
            if self.time_elapsed >= self.timer_pomodoro*60:
                self.do_timeout()

    @dbus.service.method("org.liloman.pomodoroInterface", in_signature='', out_signature='as')
    def toggle_continuous(self):
        self.continuous=not self.continuous
        return [str(self.continuous)]

    @dbus.service.method("org.liloman.pomodoroInterface", in_signature='a{ss}', out_signature='as')
    def do_start(self, dic):
        active = self.get_active_task ()
        prev_state = self.state
        self.state = "started"
        uid=dic['uuid']
        resume=dic['resume']

        #if the user wants to start the timer without any task
        if uid == '0':
            if active:
                #stop current
                active.stop()
                self.last_task_id = 0
            self.update_systray()
            return ["started with no task"]

        #if start a new one when running one already
        if active:
            if active['uuid'] != uid:
                new=self.tw.tasks.get(uuid=uid)
                active.stop()
                new.start()
                self.last_task_id = uid
                if resume == 'No':
                    self.time_elapsed = 0 
                self.update_systray()
                return ["started:"+uid]
            else:
                return ["Already started"]
        else:
            new=self.tw.tasks.get(uuid=uid)
            try:
                new.start()
            except:
                self.state = prev_state
                return ["Incorrect uuid:"+uid+" .The task can't be started"]

            if resume == 'No':
                self.time_elapsed = 0 
            self.last_task_id = uid
            self.update_systray()
            return ["started:"+uid]

   
    @dbus.service.method("org.liloman.pomodoroInterface", in_signature='s', out_signature='as')
    def do_fsm(self, event):
        fsm = "{0}-{1}".format(self.state,event)
        if fsm in self.events:
            fun = getattr(self,self.events[fsm])
            out = fun()
            self.update_systray()
            return [out]
        else:
            return ["event:"+event+" not found"]

    def close_systray(self):
        try:
            systray_bus = dbus.SessionBus()
            systray_bus = systray_bus.get_object('org.liloman.pomodoro.systray', "/systray")
            interface_systray = dbus.Interface(systray_bus, "org.liloman.pomodoro.systrayInterface")
            interface_systray.do_quit()
        except:
            # print "systray not found"
            return 

    @dbus.service.method("org.liloman.pomodoroInterface", in_signature='', out_signature='as')
    def done_current(self):
        active = self.get_active_task ()
        if not active:
            return ["no active task"]
        active.done()
        return ["ok"]

    @dbus.service.method("org.liloman.pomodoroInterface", in_signature='b', out_signature='')
    def do_quit(self,close_systray):
        #stop timer
        self.timer.cancel()
        #close systray 
        if close_systray: 
            self.close_systray()
        # Exit daemon
        loop.quit()


if __name__ == '__main__':
    dbus.mainloop.glib.DBusGMainLoop(set_as_default=True)
    loop = GObject.MainLoop()

    #if testing or with other data.location
    if len(sys.argv) >= 2:
        pomodoro = Pomodoro(sys.argv[1])
    else:
        pomodoro = Pomodoro()
    pomodoro.increment()

    task = pomodoro.get_active_task()
    if task:
        pomodoro.do_start(dbus.Dictionary({'uuid': task['uuid'], 'resume': 'No'}))
    else:
        pomodoro.do_stop()

    #run the systray
    if len(sys.argv) == 2:
            subprocess.Popen(['./systray.py', sys.argv[1]])
    elif len(sys.argv) == 1:
        subprocess.Popen(['./systray.py'])
    #dont lauch the systray when testing len(sys.argv)==3

    loop.run()
